### PR TITLE
allow skipTrash options to delete function to skip trash (hard delete)

### PR DESCRIPTION
**Pull request by ffauvel**
Allow hard delete of a record and it's associations

### CODE REVIEW

File: src/Model/Behavior/TrashBehavior.php [hunk 1]
```diff
@@ -100,6 +100,11 @@ public function implementedEvents()
      */
     public function beforeDelete(Event $event, EntityInterface $entity, ArrayObject $options)
     {
+        if (isset($options['skipTrash']) &&
```
Comment (ADmad) [inline src/Model/Behavior/TrashBehavior.php]: Using `!empty()` would be better so that it doesn't skip when `$option['skipTrash']` is set to `false`.

File: src/Model/Behavior/TrashBehavior.php [hunk 2]
```diff
@@ -100,6 +100,11 @@ public function implementedEvents()
...
+            $options['skipTrash']) {
+            return true;
```
Comment (ADmad) [inline src/Model/Behavior/TrashBehavior.php]: Should be just `return;`

File: tests/TestCase/Model/Behavior/TrashBehaviorTest.php [hunk 3]
```diff
@@ -27,11 +25,11 @@ class TrashBehaviorTest extends TestCase
      * @var array
      */
     public $fixtures = [
-        'plugin.Muffin/Trash.Articles',
-        'plugin.Muffin/Trash.Comments',
-        'plugin.Muffin/Trash.Users',
-        'plugin.Muffin/Trash.ArticlesUsers',
-        'plugin.Muffin/Trash.CompositeArticlesUsers',
+        'plugin.Muffin/Trash.articles',
```
Comment (ADmad) [inline tests/TestCase/Model/Behavior/TrashBehaviorTest.php]: Why change this? Current casing is already correct.

File: tests/TestCase/Model/Behavior/TrashBehaviorTest.php [hunk 4]
```diff
@@ -51,10 +49,7 @@ public function setUp()
             'targetForeignKey' => 'article_id',
         ]);
 
-        $this->CompositeArticlesUsers = TableRegistry::get(
-            'Muffin/Trash.CompositeArticlesUsers',
-            ['table' => 'trash_composite_articles_users']
-        );
+        $this->CompositeArticlesUsers = TableRegistry::get('Muffin/Trash.CompositeArticlesUsers', ['table' => 'trash_composite_articles_users']);
```
Comment (ADmad) [inline tests/TestCase/Model/Behavior/TrashBehaviorTest.php]: Please don't make unnecessary changes unrelated to your enhancement and add noise to the PR.

File: src/Model/Behavior/TrashBehavior.php (hunk 1 above)
Comment (ffauvel) [reply src/Model/Behavior/TrashBehavior.php]: there is a isset and a check if variable is true on the next line.
I can switch it to !empty if you prefer or I can force to use true with === true

Cakephp is using the isset approach in their CounterCacheBehavior. I was trying to follow their standards.
https://github.com/cakephp/orm/blob/master/Behavior/CounterCacheBehavior.php
```
if (isset($options['ignoreCounterCache']) && $options['ignoreCounterCache'] === true) {
    return;
}
```

File: src/Model/Behavior/TrashBehavior.php (hunk 2 above)
Comment (ffauvel) [reply src/Model/Behavior/TrashBehavior.php]: I'll switch it, I return true cause u did that at the end of the beforeDelete.

File: tests/TestCase/Model/Behavior/TrashBehaviorTest.php (hunk 3 above)
Comment (ffauvel) [reply tests/TestCase/Model/Behavior/TrashBehaviorTest.php]: My bad, my fork wasn't up to date with your latest changes. I fixed it. I'll double check my 2 tests (deletion of a record and deletion of associated record)

File: src/Model/Behavior/TrashBehavior.php (hunk 2 above)
Comment (ffauvel) [reply src/Model/Behavior/TrashBehavior.php]: Switched to !empty

File: src/Model/Behavior/TrashBehavior.php (hunk 1 above)
Comment (ffauvel) [reply src/Model/Behavior/TrashBehavior.php]: Swiched to !empty

File: src/Model/Behavior/TrashBehavior.php (hunk 2 above)
Comment (ffauvel) [reply src/Model/Behavior/TrashBehavior.php]: Replace return true; for return;

File: src/Model/Behavior/TrashBehavior.php [hunk 5]
```diff
@@ -161,6 +165,10 @@ public function trash(EntityInterface $entity, array $options = [])
      */
     public function beforeFind(Event $event, Query $query, ArrayObject $options, $primary)
     {
+        if (!empty($options['skipTrash'])) {
```
Comment (ADmad) [inline src/Model/Behavior/TrashBehavior.php]: When is the option meant for deleting checked in beforeFind() ?

Comment (ffauvel) [reply src/Model/Behavior/TrashBehavior.php]: You're right.

If cascadeCallbacks is set to true, all associations with TrashBehavior won't have their already trash records deleted.

I thought that the options was being pass over to the find function in the cascadeDelete of the associations.

Don't know how to handle that possibility.

Comment (ADmad) [reply src/Model/Behavior/TrashBehavior.php]: Delete's options are never passed to find for associated records so this check won't do anything.

File: tests/TestCase/Model/Behavior/TrashBehaviorTest.php [hunk 6]
```diff
@@ -155,6 +155,40 @@ public function testBeforeDelete()
...
+        $association->setCascadeCallbacks(true);
+
+        $article = $this->Articles->get(1);
+
+        $this->Articles->delete($article, ['skipTrash' => true]);
+
+        $comments = $this->Comments->find('withTrashed')
+            ->where(['article_id' => 1]);
+
+        $this->assertCount(0, $comments);
```
Comment (ADmad) [inline tests/TestCase/Model/Behavior/TrashBehaviorTest.php]: `$comments` is an instance of `Query()` here, that's why your test fails. You need to do `count()`.
